people-and-planet-ai/timeseries-classification/trainer.py

```diff
@@ -0,0 +1,272 @@
+        class GeoPoint(keras.layers.Layer):
+            def call(self: Any, latlon: Tuple[tf.Tensor, tf.Tensor]) -> tf.Tensor:
+                lat, lon = latlon
+                x = tf.cos(lon) * tf.sin(lat)
+                y = tf.sin(lon) * tf.sin(lat)
+                z = tf.cos(lat)
+                return tf.concat([x, y, z], axis=-1)
+
+        lat_lon_input_layers = (input_layers[lat_name], input_layers[lon_name])
+        return GeoPoint(name=f"{lat_name}_{lon_name}")(lat_lon_input_layers)
+
+    def sequential_layers(
+        first_layer: keras.layers.Layer, *layers: keras.layers.Layer
+    ) -> keras.layers.Layer:
+        return reduce(lambda layer, result: result(layer), layers, first_layer)
+
+    preprocessed_inputs = [
+        normalize("distance_from_port"),
+        normalize("speed"),
+        direction("course"),
+        geo_point("lat", "lon"),
+    ]
+
+    output_layers = {
+        "is_fishing": sequential_layers(
+            keras.layers.concatenate(preprocessed_inputs, name="deep_layers"),
+            keras.layers.Conv1D(
+                filters=32,
+                kernel_size=PADDING + 1,
+                data_format="channels_last",
+                activation="relu",
+            ),
+            keras.layers.Dense(16, activation="relu"),
+            keras.layers.Dense(1, activation="sigmoid", name="is_fishing"),
+        )
+    }
+    return keras.Model(input_layers, output_layers)
+
+
+def run(
+    train_data_dir: str,
+    eval_data_dir: str,
+    train_steps: int,
+    eval_steps: int,
+    batch_size: int,
+    model_dir: str,
+    checkpoint_dir: str,
+    tensorboard_dir: str,
+) -> None:
+
+    # For this sample we are using a mirrored distribution strategy,
+    # which consists of a single machine with multiple GPUs.
+    #   https://blog.tensorflow.org/2020/12/getting-started-with-distributed-tensorflow-on-gcp.html
+    distributed_strategy = tf.distribute.MirroredStrategy()
+
+    # Create the training and evaluation datasets from the TFRecord files.
+    logging.info("Creating datasets")
+    total_batch_size = batch_size * distributed_strategy.num_replicas_in_sync
+    train_dataset = create_dataset(train_data_dir, total_batch_size)
+    eval_dataset = create_dataset(eval_data_dir, total_batch_size)
+
+    # Create and compile the model inside the distribution strategy scope.
+    with distributed_strategy.scope():
+        logging.info("Creating the model")
+        model = create_model(train_dataset)
+
+        logging.info("Compiling the model")
+        model.compile(
+            optimizer="adam",  # https://keras.io/api/optimizers
+            loss={"is_fishing": "binary_crossentropy"},  # https://keras.io/api/losses
+            metrics={"is_fishing": ["accuracy"]},  # https://keras.io/api/metrics
+        )
+
+    # Train the model.
+    logging.info("Training the model")
+    model.fit(
+        train_dataset.repeat(),
```

Comment:
My recommendation would be to pass in the number of epochs instead of the steps per epoch. Keras will calculate the correct number of steps (num samples/batch size). You don't need to repeat the dataset in this case either. When you factor in distributed training and start scaling the batch size, your steps per epoch will change as well since you're now seeing more data in each step. IMO passing the epoch and letting keras do the calculation is just less complexity.

Reply:
Thank you! This is awesome. I had some trouble, but in the end I still needed to pass `eval_steps` for validation, even if the training used epochs. If I didn't pass `eval_steps` I got an error of having "empty" data points. If you know of a better way to do it, please let me know! But having it this way, at least tests are passing again.